### DISCUSSION TITLE
How to have two validation loaders in validation loop

### DISCUSSION THREAD

**Discussion by LanceKnight** [Lightning-AI/pytorch-lightning#10364] [General]
Hello, I'm training with dropouts. However, when I compared loss in training and validation, they're not really comparable since the former has dropout but the later does not. So I'm trying to run the training dataset in the validation loop as well so I can get a training loss without dropout. But I'm not sure how to pass both a training and validation data loaders to the validation loop?

**Comment by i008** [comment]
You can return two loaders in your val_dataloader:

```
def val_dataloader(self):
    return train_load, val_load
```

pl will go trough both of them and the data will be available for you. Remember to accept the dataloader_ix in you validation_step.
And in your validation_epoch_end the outputs/results will contain a list with a len == to  the amount of data_loaders you specified.

**Reply by LanceKnight** [reply]
I tried this. But I guess the way I accept it is wrong...    

def validation_step(self, batched_data, batch_idx):
        print(f'\nactual validation step running')
        print(f'batch_data:{batched_data[0]}')

**Reply by i008** [reply]
You should also define a dataloader_ix in your functions args.
Remember whenever you have problems such as this, or you dont understand whats really going on under the hood you can do:

```
def validation_step(*args, **kwargs):
    import pdb; pdb.set_trace()
    ....
```

And inspect whatever is going on by looking how this functions was called.

**Reply by LanceKnight** [reply]
I see. Thank you for your help @i008 !

**Comment by ananthsub** [comment]
Do you have an example input you want to compare the difference on? You could extract your nn.Module from your LightningModule and compare the difference between training mode vs eval mode like so:

```
input = ...
my_model = MyModel()
out_train = my_model(input)
eval_model = my_model.eval()
with torch.no_grad():
    out_eval = eval_model(input)

# compare difference between out_train and out_eval
```

https://pytorch.org/docs/stable/notes/autograd.html#locally-disable-grad-doc

**Reply by LanceKnight** [reply]
I tried this. But CUDA is out of memory... why would cuda out of memory with this?

**Reply by tchaton** [reply]
Dear @LanceKnight, you want both dataloader to be fetched within the same batch, you can use PyTorch Lightning CombinedLoader.

```py
from pytorch_lightning.trainer.supporters import CombinedLoader

class DataModule(LightningDatataModule):

     def val_dataloader(self):
              return CombinedLoader([dl_1, dl_2])
```

**Reply by LanceKnight** [reply]
Gotcha. Thanks! @tchaton
